### ISSUE TITLE
More flexible snapshot schedule

### ISSUE THREAD

**Issue by simondeziel** [lxc/lxd#7237]
It would be nice to have `snapshots.schedule` support the `@reboot` keyword. Such schedule would mean a snapshot would be taken just before starting the instance.

It would also be nice for the `snapshots.expiry` to support a syntax like `3@reboot` to keep the past 3 snapshots taken before startups. This would likely require the snapshots name to have a `snapshots.pattern` including the `reboot` keyword like `snap_reboot%d` maybe?

Supporting other keywords like `@daily`, `@hourly`, etc would be convenient too and would fit nicely with an extension of the `snapshots.schedule` and `snapshots.expiry` parameters to accept comma delimited lists.

This would permit one to ask for 3 `@reboot`, 24 `@hourly` and 7 `@daily` snapshots to be kept with a simple syntax:

```
snapshots.schedule: @reboot, @hourly, @daily
snapshots.expiry:  3@reboot, 24@hourly, 7@daily
```

**Comment by stgraber** [comment]
I'm not sure how doable the `snapshots.expiry` piece is. The way this works in LXD is that we write the ExpiresAt value at the time the snapshot is made, so we can't really have such, keep 3 of this type kind of logic in there.

The changes to scheduling logic should be fine though.

For expiry, if you use specific naming patterns, writing a script which handles cleanup through the API should be pretty simple and would avoid having to add a lot of potentially complex logic in LXD for that.

**Comment by stgraber** [comment]
So I think we should:
 - Allow for multiple comma separated schedules in `snapshots.schedule` for both instances and custom storage volumes
 - Allow for the common cron shortcuts:
   - `@hourly` => `0 * * * *`
   - `@daily` and `@midnight` => `0 0 * * *`
   - `@weekly` => `0 0 * * 0`
   - `@monthly` => `0 0 1 * *`
   - `@annually` and `@yearly` => `0 0 1 1 *`
 - Only for instances, also allow `@reboot` which will trigger a snapshot right before startup

**Comment by simondeziel** [comment]
> I'm not sure how doable the `snapshots.expiry` piece is. The way this works in LXD is that we write the ExpiresAt value at the time the snapshot is made, so we can't really have such, keep 3 of this type kind of logic in there.

What if the snapshot type (`@reboot`, `@daily`, etc) is added to the DB when it is created?

**Comment by stgraber** [comment]
Yeah, that's possible, though we'd probably want to see more requests for this before considering adding a field to both DB and API.

**Comment by simondeziel** [comment]
Understood. Having the support for `@reboot` snapshots is the most important feature here. This is especially handy with VMs that are harder to snapshot live.

**Comment by zhenr667** [comment]
Hey, I'm a student at UT and my team would like to claim this issue for a class, if that's ok!

**Comment by stgraber** [comment]
Thanks, assigned it to you.

https://github.com/lxc/lxd/issues/7237#issuecomment-617495306 instead of the issue description is what we should really be aiming for here.

**Comment by stgraber** [comment]
@zhenr667  Do you still intend to work on this or should we un-assign so someone else can have a go at it?

**Comment by zhenr667** [comment]
@stgraber I don't think I can work on this, sorry for the trouble.

**Comment by my5t3ry** [comment]
Hello, 

if you have no objections i would start working on this feature, but can predict if i'm able to implement it. :)

**Comment by stgraber** [comment]
Sure, assigned it to you!

**Comment by my5t3ry** [comment]
ok great, it wil take a couple of days till i find time for it because of wage labour tickets :)

**Comment by sqllyw** [comment]
I'll assume the snapshot taking is for the container itself, but sometimes a zfs dataset in the host was attached:

sudo lxc config device add centos6-i386 dataset disk source=/mypool/database/data path=/data

will this be included?

or is there a possibility to run a user defined script at that point? thanks

**Comment by stgraber** [comment]
Yeah, that would only affect the instance itself, though the same logic could be extended to custom storage volumes managed by LXD.

Completely external disks as you describe will not get snapshotted. We have some plans to introduce some kind of webhooks down the line which you could use to trigger such extra steps. In the meantime, you could write a small daemon that reacts to lifecycle events sent by LXD as we do emit such events whenever we make a snapshot.

**Comment by sqllyw** [comment]
lifecycle event should be enough, any documentation on this? thanks

**Comment by stgraber** [comment]
You can see them by doing `lxc monitor --type=lifecycle`.
They get sent over `/1.0/events?type=lifecycle` which is a websocket endpoint that you can listen on.

If using Go to connect to it, you'd use `c.GetEvents()` to get an EventListener on which you'd then call `AddHandler([]string{"lifecycle"}, func)` with `func` being a function you wrote that will then get called every time a lifecycle event is received and can inspect it to decide whether it should do something.

You can see me doing something like that with my BGP server here: https://github.com/stgraber/lxd-bgp/blob/master/cmd/lxd-bgp/main.go#L97

In this case, I'm listening for any instance start/stop events as well as any kind of instance reconfiguration.
